internal/pkg/log/assert.go

```diff
@@ -0,0 +1,88 @@
+package log
+
+import (
+	"bufio"
+	"strings"
+
+	"github.com/keboola/go-utils/pkg/wildcards"
+	"github.com/stretchr/testify/assert"
+
+	"github.com/keboola/keboola-as-code/internal/pkg/encoding/json"
+	"github.com/keboola/keboola-as-code/internal/pkg/utils/errors"
+)
+
+// CompareJSONMessages checks that expected json messages appear in actual in the same order.
+// Actual string may have extra messages and the rest may have extra fields. String values are compared using wildcards.
+// Returns nil if the expectations are met or an error with the first unmatched expected line and all remaining actual lines.
+func CompareJSONMessages(expected string, actual string) error {
+	expectedScanner := bufio.NewScanner(strings.NewReader(strings.Trim(expected, "\n")))
+	actualScanner := bufio.NewScanner(strings.NewReader(strings.Trim(actual, "\n")))
+
+	for expectedScanner.Scan() {
+		expectedMessage := expectedScanner.Text()
+		var expectedMessageData map[string]any
+		err := json.DecodeString(expectedMessage, &expectedMessageData)
+		if err != nil {
+			return errors.Wrapf(err, "expected string contains invalid json:\n%s", expectedMessage)
+		}
+
+		actualMessages := ""
+		messageFound := false
+		for actualScanner.Scan() {
+			actualMessage := actualScanner.Text()
+			actualMessages += actualMessage + "\n"
+			var actualMessageData map[string]any
+			err := json.DecodeString(actualMessage, &actualMessageData)
+			if err != nil {
+				return errors.Wrapf(err, "actual string contains invalid json:\n%s", actualMessage)
+			}
+
+			messageFound = true
+			for key, value := range expectedMessageData {
+				actualValue, ok := actualMessageData[key]
+				if !ok || !valueMatches(value, actualValue) {
+					messageFound = false
+					break
+				}
+			}
+
+			if messageFound {
+				break
+			}
+		}
+
+		if !messageFound {
+			return errors.Errorf(
+				"Expected:\n-----\n%s\n-----\nActual:\n-----\n%s",
+				expectedMessage,
+				strings.TrimRight(actualMessages, "\n"),
+			)
+		}
+	}
+
+	return nil
+}
+
+func valueMatches(value any, actualValue any) bool {
+	if expectedString, ok := value.(string); ok {
+		if actualString, ok := actualValue.(string); ok {
+			err := wildcards.Compare(expectedString, actualString)
+			return err == nil
+		}
+
+		return false
+	}
+
+	return actualValue == value
```

Comment:
As I look at it a second time, I found one more case when it doesn't work (sorry I didn't notice it earlier):

```
		{
			name:     "match not scalar fields",
			expected: `{"level":"info","message":"Info msg","array":["foo","bar"],"object":{"foo":"bar"}}`,
			actual:   `{"level":"info","message":"Info msg","array":["foo","bar"],"object":{"foo":"bar"}}`,
		},
		{
			name:     "match not scalar fields - fail",
			expected: `{"level":"info","message":"Info msg","array":["123","456"],"object":{"foo":"bar"}}`,
			actual:   `{"level":"info","message":"Info msg","array":["foo","bar"],"object":{"foo":"bar"}}`,
			err: errors.Errorf(
				"Expected:\n-----\n%s\n-----\nActual:\n-----\n%s",
				`{"level":"info","message":"Info msg","array":["123","456"],"object":{"foo":"bar"}}`,
				`{"level":"info","message":"Info msg","array":["foo","bar"],"object":{"foo":"bar"}}`,
			),
		},	
```

![image](https://github.com/keboola/keboola-as-code/assets/19371734/ec3b4c20-af0c-479a-a000-d8290d7b6e9a)

------------------

:arrow_right: Please, use `return reflect.DeepEqual(actualValue, value)` here.

Reply:
Thanks, there may by a `tags` slice, for example, and it is supported by `ctxattr`: https://github.com/keboola/keboola-as-code/blob/c10891eb0402162d1abab70b952cb1ca627bba49/internal/pkg/service/common/ctxattr/ctxattr.go#L82-L83